### PR TITLE
Don't allow detachedConfigurations to extend

### PR DESCRIPTION
**Pull request by tresat**
Addresses: https://github.com/gradle/gradle/issues/6881

By deprecating this behavior.

## Problem with failing fast

Failing-fast by throwing an exception when you attempt to have a detached configuration extend another configuration is the goal, but can't be done yet.

Doing this fails our `PalantirConsistentVersionsPluginSmokeTest`, as that plugin [attempts](https://github.com/palantir/gradle-consistent-versions/blob/d09653ef15e904ab881da547149ec0bb7f29ad30/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L123) to [wire all configurations to extends a new root configuration](https://github.com/palantir/gradle-consistent-versions/blob/d09653ef15e904ab881da547149ec0bb7f29ad30/src/main/java/com/palantir/gradle/versions/VersionsPropsPlugin.java#L213), which now causes an exception.

The detached configuration in question is a copy:
```
        Failures:
           - Could not resolve root project :.
           - configuration ':compileClasspathCopy' is a detached configuration and thus cannot extend configuration(s): 'rootConfiguration'
```

Because the plugin copies configurations using [copyRecursive](https://github.com/palantir/gradle-consistent-versions/blob/d09653ef15e904ab881da547149ec0bb7f29ad30/src/main/java/com/palantir/gradle/versions/VersionsLockPlugin.java#L640), which calls `DefaultConfiguration#copyAsDetached`.

So this will have to be a deprecation for now.

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 1]
```diff
@@ -217,6 +217,13 @@ The existing method does not behave as specified by its javadoc, and we do not p
 
 Potential migrations include using `Object.equals(Object)` directly, or comparing the fields of dependencies manually.
 
+[[detached_configurations_cannot_extend]]
+==== Detached Configurations Should Not Use ExtendFrom
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:221]:
```suggestion
==== Detached Configurations should not use `extendFrom`
```

Seems like we don't capitalize other titles

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 2]
```diff
@@ -217,6 +217,13 @@ The existing method does not behave as specified by its javadoc, and we do not p
...
+
+link:{javadocPath}/org/gradle/api/artifacts/ConfigurationContainer.html#detachedConfiguration(org.gradle.api.artifacts.Dependency...)[Detached configurations] should not ever extend other configurations using link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#extendsFrom(org.gradle.api.artifacts.Configuration...)[extendsFrom].
+
+This behavior has been deprecated and will be removed in Gradle 9.0.
```
Comment (jvandort) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:225]:
```suggestion
This behavior has been deprecated and will become an error in Gradle 9.0.
```

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 3]
```diff
@@ -1855,6 +1858,19 @@ public ConfigurationRole getRoleAtCreation() {
         return roleAtCreation;
     }
 
+    private void assertNotDetachedExtensionDoingExtending(Iterable<Configuration> extendsFrom) {
+        if (isDetachedConfiguration()) {
+            String summarizedExtensionTargets = StreamSupport.stream(extendsFrom.spliterator(), false)
+                .map(c -> "'" + c.getName() + "'")
+                .collect(Collectors.joining(", "));
+            DeprecationLogger.deprecateAction(String.format("Calling extendsFrom on %s", this.getDisplayName()))
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1866]:
```suggestion
            DeprecationLogger.deprecateAction("Calling extendsFrom on a detached configuration")
```

I think we should put the thing thats being deprecated here and the configuration in question in the context. We're deprecating extending any detached configuration, not specifically this one detached configuration

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 4]
```diff
@@ -371,6 +372,7 @@ public Set<Configuration> getExtendsFrom() {
     @Override
     public Configuration setExtendsFrom(Iterable<Configuration> extendsFrom) {
         validateMutation(MutationType.HIERARCHY);
+        assertNotDetachedExtensionDoingExtending(extendsFrom);
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:375]: We also call `extendsFrom(Configuration...)` from this method.

We should either remove this line and allow the `extendsFrom(configuration)` to report the deprecated action, or wrap the `extendsFrom(configuration)` line below in a `DeprecationLogger.whileDisabled`

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java [hunk 5]
```diff
@@ -1855,6 +1858,19 @@ public ConfigurationRole getRoleAtCreation() {
         return roleAtCreation;
     }
 
+    private void assertNotDetachedExtensionDoingExtending(Iterable<Configuration> extendsFrom) {
+        if (isDetachedConfiguration()) {
+            String summarizedExtensionTargets = StreamSupport.stream(extendsFrom.spliterator(), false)
+                .map(c -> "'" + c.getName() + "'")
+                .collect(Collectors.joining(", "));
+            DeprecationLogger.deprecateAction(String.format("Calling extendsFrom on %s", this.getDisplayName()))
+                .withContext(String.format("This is a detached configurations and can not extend other configurations, in this case: %s.", summarizedExtensionTargets))
```
Comment (jvandort) [inline platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1867]:
```suggestion
                .withContext(String.format("Detached %s should not extend configuration(s): %s.", this.getDisplayName(), summarizedExtensionTargets))
```

This wording seemed a little odd to me

**Review by jvandort — APPROVED** [review]

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 3 above)
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:1866]: Can't do this, as the `displayName` contains the `configuration:` part as a prefix.  

Detached configurations are all named `detachedConfigurationX`, and the context bit of the message says "This is a detached configuration and can not extend other configurations" so it should be clear.

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 5 above)
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java]: Adjusted the wording a little

File: platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java (hunk 4 above)
Comment (tresat) [reply platforms/software/dependency-management/src/main/java/org/gradle/api/internal/artifacts/configurations/DefaultConfiguration.java:375]: It shouldn't matter, as (currently) the deprecation logger will automatically de-duplicate and only log the message once and (in Gradle 9.0) the exception will be thrown by the call into the assertion method, precluding the second call.  

Just as we call `validateMutation` twice, we should call this assertion method twice, to be consistent with the pattern.

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

**Comment by tresat** [comment]
@bot-gradle test and merge

File: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc [hunk 6]
```diff
@@ -217,6 +217,13 @@ The existing method does not behave as specified by its javadoc, and we do not p
 
 Potential migrations include using `Object.equals(Object)` directly, or comparing the fields of dependencies manually.
 
+[[detached_configurations_cannot_extend]]
+==== Detached Configurations should not use `extendFrom`
+
+link:{javadocPath}/org/gradle/api/artifacts/ConfigurationContainer.html#detachedConfiguration(org.gradle.api.artifacts.Dependency...)[Detached configurations] should not ever extend other configurations using link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#extendsFrom(org.gradle.api.artifacts.Configuration...)[extendsFrom].
```
Comment (lkasso) [inline platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc:223]:
```suggestion
link:{javadocPath}/org/gradle/api/artifacts/ConfigurationContainer.html#detachedConfiguration(org.gradle.api.artifacts.Dependency...)[Detached configurations] should not extend other configurations using link:{javadocPath}/org/gradle/api/artifacts/Configuration.html#extendsFrom(org.gradle.api.artifacts.Configuration...)[extendsFrom].
```

**Comment by tresat** [comment]
@bot-gradle test and merge

**Pull request merged**